crates/bevy_reflect/src/struct_trait.rs

```diff
@@ -337,20 +337,29 @@ impl Struct for DynamicStruct {
     }
 }
 
-// SAFE: any and any_mut both return self
-unsafe impl Reflect for DynamicStruct {
+impl Reflect for DynamicStruct {
     #[inline]
     fn type_name(&self) -> &str {
         &self.name
     }
 
+    #[inline]
+    fn type_id(&self) -> TypeId {
+        TypeId::of::<Self>()
```

Comment:
I'm curious if anyone else finds this counter-intuitive, since `Dynamic***` structs return their type name dynamically. One might assume they do the same for `TypeId`.

But at the same time, we can't _not_ do this since then the `is()` methods would incorrectly return true (unless that's intended?)

Reply:
This makes sense to me; I would be very surprised if dynamic types did anything else.

Reply:
This is a weird one actually, and I didn't realize it while putting this together. I *would* have dynamic structs return their represented type's TypeId, but what is the case for if they were constructed without that information?
They can't be downcasted to anything, though, so it would cause downcast to panic if you gave it the "correct" type. I think returning an option like Alice mentioned below is a good idea

Reply:
This is discussed in [rfc56](https://github.com/bevyengine/rfcs/pull/56). It stagnated for a month, but I expect to pick up steam again.